src/charm.py

```diff
@@ -15,89 +15,89 @@
+        self.framework.observe(self.on.upgrade_charm, self._on_upgrade_charm)
+        self.framework.observe(self.on.start, self._on_start)
+        self.framework.observe(self.on.stop, self._on_stop)
+        self.framework.observe(self.on.website_relation_changed, self._on_website_relation_changed)
+        self._grafana_agent = COSAgentProvider(
+            self,
+            metrics_endpoints=[
+                {"path": "/metrics", "port": METRICS_PORT},
+            ],
+            metrics_rules_dir="./src/prometheus_alert_rules",
+            dashboard_dirs=["./src/grafana_dashboards"],
+        )
+        self.pollen = PollenService()
+        self.hostname = self.model.get_binding("website").network.bind_address
+        self._charm_state = CharmState.from_charm(self)
+
+    def _on_website_relation_changed(self, event: ops.RelationChangedEvent):
+        """Handle website-relation-changed.
+
+        Args:
+            event: Event triggering the website relation joined handler.
+        """
+        event.relation.data[self.unit].update(self._charm_state.website)
+
+    def _on_install(self, event: ops.InstallEvent):
+        """Handle install.
 
-    def _on_httpbin_pebble_ready(self, event: ops.PebbleReadyEvent):
-        """Define and start a workload using the Pebble API.
+        Args:
+            event: Event triggering the install handler.
+        """
+        self.unit.status = MaintenanceStatus("Installing dependencies")
+        self.pollen.prepare(str(self.unit.name).replace("/", "-"), self._charm_state)
 
-        Change this example to suit your needs. You'll need to specify the right entrypoint and
-        environment configuration for your specific workload.
+    def _on_upgrade_charm(self, event: ops.UpgradeCharmEvent):
+        """Handle upgrade-charm.
 
-        Learn more about interacting with Pebble at at https://juju.is/docs/sdk/pebble.
+        Args:
+            event: Event triggering the upgrade-charm handler.
         """
-        # Get a reference the container attribute on the PebbleReadyEvent
-        container = event.workload
-        # Add initial Pebble config layer using the Pebble API
-        container.add_layer("httpbin", self._pebble_layer, combine=True)
-        # Make Pebble reevaluate its plan, ensuring any services are started if enabled.
-        container.replan()
-        # Learn more about statuses in the SDK docs:
-        # https://juju.is/docs/sdk/constructs#heading--statuses
-        self.unit.status = ops.ActiveStatus()
-
-    def _on_config_changed(self, event: ops.ConfigChangedEvent):
-        """Handle changed configuration.
-
-        Change this example to suit your needs. If you don't need to handle config, you can remove
-        this method.
-
-        Learn more about config at https://juju.is/docs/sdk/config
+        self.unit.status = MaintenanceStatus("Upgrading dependencies")
+        self.pollen.prepare(str(self.unit.name).replace("/", "-"), self._charm_state)
```

Comment:
Looks like `str(self.unit.name).replace("/", "-")` is repeated a few time, perhaps worth a function?

Reply:
It is only repeated two times, I think it would be better to leave it as it is